### PR TITLE
Initial support for building on FreeBSD

### LINKED ISSUE #12787: compilation fails on FreeBSD 13.2 RELEASE
- [x ] Checked for duplicates

### Describe the bug

segfault in build process. 
### Expected behavior

build process completes.

### To Reproduce

standard cmake build process, as described in the guide.

### Setup

ROOT v6-26-10 (@4dddea35)
FreeBSD 13.2 RELEASE (clang 14.0.5), required dependencies installed from ports

### Additional context

a port exists v6-28-02 (development/root), installs in binary form (so somebody got it to compile, did not have too much time to test it thoroughly). In my case, however, the port also does not compile, craps out somewhere else. plus it hauls in too many dependencies so it is less than ideal.
I imagine getting it to work on FreeBSD would not end at fixing this issue here. The fbsd port maintainer has some patches which might help, but I have also seen that e.g. procfs is deprecated in new releases and libprocstat has to be used, etc, etc.
With that being said, it would be extremely nice to get ROOT to work on FreeBSD...
thanks,

```
[ 52%] Built target rootcling_stage1
[ 52%] Generating G__Core.cxx, ../lib/Core.pcm
 #0 0x000000000446e3d5 llvm::sys::PrintStackTrace(llvm::raw_ostream&) Signals.cpp:0:0
 #1 0x000000000446c646 llvm::sys::RunSignalHandlers() Signals.cpp:0:0
 #2 0x000000000446eba6 SignalHandler(int) Signals.cpp:0:0
 #3 0x0000000826961b60 (/lib/libthr.so.3+0x19b60)
 #4 0x000000082696111f (/lib/libthr.so.3+0x1911f)
 #5 0x00007ffffffff8a3 ([vdso]+0x2d3)
 #6 0x000000082da5f7b4 strlen (/lib/libc.so.7+0x1587b4)
 #7 0x00000000016482c9 std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::basic_string<std::nullptr_t>(char const*) rootcling_stage1.cxx:0:0
 #8 0x0000000001648213 GetEtcDir() rootcling_stage1.cxx:0:0
 #9 0x00000000016b2792 RootClingMain(int, char**, bool) (/home/mkrzewic/src/root_build/core/rootcling_stage1/src/rootcling_stage1+0x16b2792)
#10 0x00000000016c32d5 ROOT_rootcling_Driver (/home/mkrzewic/src/root_build/core/rootcling_stage1/src/rootcling_stage1+0x16c32d5)
Stack dump:
0.	Program arguments: /home/mkrzewic/src/root_build/core/rootcling_stage1/src/rootcling_stage1 -v2 -f G__Core.cxx -cxxmodule -s /home/mkrzewic/src/root_build/lib/libCore.so -excludePath /home/mkrzewic/src/root_src -excludePath /home/mkrzewic/src/root_build/ginclude -excludePath /home/mkrzewic/src/root_build/externals -excludePath /home/mkrzewic/src/root_build/builtins -writeEmptyRootPCM -m _Builtin_intrinsics -mByproduct _Builtin_intrinsics -mByproduct ROOT_Foundation_Stage1_NoRTTI -mByproduct ROOT_Foundation_C -mByproduct ROOT_Rtypes -D__STDC_NO_COMPLEX__ -D__COMPLEX_H__ -D_COMPLEX_H -D__CLANG_STDATOMIC_H -compilerI/usr/include/c++/v1 -compilerI/usr/lib/clang/14.0.5/include -compilerI/usr/include -compilerI/usr/lib/clang/14.0.5/include -compilerI/usr/include -I/home/mkrzewic/src/root_build/include -I/home/mkrzewic/src/root_build/ginclude -I/home/mkrzewic/src/root_src/core/base/inc -I/home/mkrzewic/src/root_src/core/foundation/inc -I/home/mkrzewic/src/root_src/core/cont/inc -I/home/mkrzewic/src/root_src/core/gui/inc -I/home/mkrzewic/src/root_src/core/meta/inc -I/home/mkrzewic/src/root_src/core/clib/inc -I/home/mkrzewic/src/root_src/core/rint/inc -I/home/mkrzewic/src/root_src/core/zip/inc -I/home/mkrzewic/src/root_src/core/thread/inc -I/home/mkrzewic/src/root_src/core/textinput/inc -I/home/mkrzewic/src/root_src/core/clingutils/inc -I/home/mkrzewic/src/root_src/core/base/v7/inc -I/home/mkrzewic/src/root_src/core/foundation/v7/inc -I/home/mkrzewic/src/root_src/core/unix/inc -I/usr/local/include -I/home/mkrzewic/src/root_src/builtins/xxhash -I/usr/include ROOT/TErrorDefaultHandler.hxx ROOT/TSequentialExecutor.hxx ROOT/StringConv.hxx Buttons.h Bytes.h Byteswap.h KeySymbols.h MessageTypes.h Riostream.h Rtypes.h TApplication.h TAtt3D.h TAttAxis.h TAttBBox2D.h TAttBBox.h TAttFill.h TAttLine.h TAttMarker.h TAttPad.h TAttText.h TBase64.h TBenchmark.h TBuffer3D.h TBuffer3DTypes.h TBuffer.h TColor.h TColorGradient.h TDatime.h TDirectory.h TEnv.h TException.h TExec.h TFileCollection.h TFileInfo.h TFolder.h TInetAddress.h TMacro.h TMathBase.h TMD5.h TMemberInspector.h TMessageHandler.h TNamed.h TNotifyLink.h TObject.h TObjString.h TParameter.h TPluginManager.h TPoint.h TPRegexp.h TProcessID.h TProcessUUID.h TQClass.h TQCommand.h TQConnection.h TQObject.h TRedirectOutputGuard.h TRefCnt.h TRef.h TRegexp.h TRemoteObject.h TROOT.h TRootIOCtor.h TStopwatch.h TStorage.h TString.h TStringLong.h TStyle.h TSysEvtHandler.h TSystemDirectory.h TSystemFile.h TSystem.h TTask.h TThreadSlots.h TTime.h TTimer.h TTimeStamp.h TUri.h TUrl.h TUUID.h TVersionCheck.h TVirtualAuth.h TVirtualFFT.h TVirtualGL.h TVirtualMonitoring.h TVirtualMutex.h TVirtualPadEditor.h TVirtualPad.h TVirtualPadPainter.h TVirtualPerfStats.h TVirtualPS.h TVirtualQConnection.h TVirtualRWMutex.h TVirtualTableInterface.h TVirtualViewer3D.h TVirtualX.h ROOT/RDirectoryEntry.hxx ROOT/RIndexIter.hxx strlcpy.h snprintf.h strtok.h ROOT/TSeq.hxx TArrayC.h TArrayD.h TArrayF.h TArray.h TArrayI.h TArrayL64.h TArrayL.h TArrayS.h TBits.h TBtree.h TClassTable.h TClonesArray.h TCollection.h TCollectionProxyInfo.h TExMap.h THashList.h THashTable.h TIterator.h TList.h TMap.h TObjArray.h TObjectTable.h TOrdCollection.h TRefArray.h TRefTable.h TSeqCollection.h TSortedList.h TVirtualCollectionProxy.h ESTLType.h RStringView.h TClassEdit.h TError.h ROOT/RLogger.hxx ROOT/RNotFn.hxx ROOT/RRangeCast.hxx ROOT/RSpan.hxx ROOT/RStringView.hxx ROOT/StringUtils.hxx ROOT/span.hxx ROOT/TypeTraits.hxx TUnixSystem.h root_std_complex.h GuiTypes.h TApplicationImp.h TBrowser.h TBrowserImp.h TCanvasImp.h TClassMenuItem.h TContextMenu.h TContextMenuImp.h TControlBarImp.h TGuiFactory.h TInspectorImp.h TObjectSpy.h TToggleGroup.h TToggle.h TBaseClass.h TClassGenerator.h TClass.h TClassRef.h TClassStreamer.h TDataMember.h TDataType.h TDictAttributeMap.h TDictionary.h TEnumConstant.h TEnum.h TFileMergeInfo.h TFunction.h TFunctionTemplate.h TGenericClassInfo.h TGlobal.h TInterpreter.h TInterpreterValue.h TIsAProxy.h TListOfDataMembers.h TListOfEnums.h TListOfEnumsWithLock.h TListOfFunctions.h TListOfFunctionTemplates.h TMemberStreamer.h TMethodArg.h TMethodCall.h TMethod.h TProtoClass.h TRealData.h TSchemaHelper.h TSchemaRule.h TSchemaRuleSet.h TStatusBitsChecker.h TStreamerElement.h TStreamer.h TVirtualIsAProxy.h TVirtualRefProxy.h TVirtualStreamerInfo.h TVirtualArray.h TVirtualObject.h Getline.h /home/mkrzewic/src/root_src/core/base/inc/LinkDef.h 
Segmentation fault
gmake[2]: *** [core/CMakeFiles/G__Core.dir/build.make:474: core/G__Core.cxx] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:32925: core/CMakeFiles/G__Core.dir/all] Error 2
gmake: *** [Makefile:156: all] Error 2
```

### PR DESCRIPTION
**Pull request by mkrzewic**
# This Pull request:
This is an initial attempt at getting this thing building on modern FreeBSD (i.e. without /proc)
some discussion was going on in issue #12787 and @hahnjo suggested a broader review and discussion in a PR.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #12787

### CODE REVIEW

**Comment by bellenot** [comment]
@phsft-bot build

File: core/unix/CMakeLists.txt
Comment (Axel-Naumann) [inline core/unix/CMakeLists.txt:20]: Indentation should be 2 spaces, no tabs.

File: interpreter/cling/include/cling/libc.modulemap
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/libc.modulemap:73]: @vgvassilev FYI

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:59]:
```suggestion
#   define ZFS
#   define defined_ZFS_for_libprocstat
# endif
# include <libprocstat.h>
# ifdef defined_ZFS_for_libprocstat
#   undef ZFS
#   undef defined_ZFS_for_libprocstat
# endif

# include <libutil.h>
#endif
```

**Comment by hahnjo** [comment]
Thanks @mkrzewic! We kicked off testing to make sure it doesn't break other platforms. I will take a closer look tomorrow :wink:

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:46]: Could you be more precise in the comment in what this works around?

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
This is fantastic, thank you so much!

Just suggesting some small wording (`thisisabloodyhackZFS`) / indentation changes. I'm curious to see how the CI will deal with the textual module!

File: interpreter/cling/include/cling/libc.modulemap
Comment (vgvassilev) [reply interpreter/cling/include/cling/libc.modulemap:73]: That’s probably something to avoid. What’s the underlying issue with that header file?

Comment (mkrzewic) [reply interpreter/cling/include/cling/libc.modulemap:73]: it goes like this:

[ 73%] Building CXX object core/metacling/src/CMakeFiles/MetaCling.dir/TClingMethodInfo.cxx.o
While building module 'Core':
While building module 'std' imported from input_line_1:1:
While building module 'xlocale.h' imported from /usr/include/c++/v1/__locale:39:
In file included from <module-includes>:2:
/usr/include/xlocale.h:37:10: fatal error: cyclic dependency in module 'std': std -> xlocale.h -> std
#include <locale.h>
^
While building module 'Core':
While building module 'std' imported from input_line_1:1:
In file included from <module-includes>:31:
In file included from /usr/include/c++/v1/complex.h:28:
In file included from /usr/include/c++/v1/ccomplex:20:
In file included from /usr/include/c++/v1/complex:242:
In file included from /usr/include/c++/v1/sstream:184:
In file included from /usr/include/c++/v1/istream:163:
In file included from /usr/include/c++/v1/ostream:139:
In file included from /usr/include/c++/v1/ios:214:
/usr/include/c++/v1/__locale:39:11: fatal error: could not build module 'xlocale.h'
# include <xlocale.h>
~~~~~~~~^
[ 73%] Building CXX object core/metacling/src/CMakeFiles/MetaCling.dir/TClingRdictModuleFileExtension.cxx.o
input_line_1:1:10: fatal error: could not build module 'std'
#include <new>
~~~~~~~~^
Warning in cling::IncrementalParser::CheckABICompatibility():
Failed to extract C++ standard library version.
Warning in cling::IncrementalParser::CheckABICompatibility():
Possible C++ standard library mismatch, compiled with _LIBCPP_ABI_VERSION '1'
Extraction of runtime standard library version was: ''

File: core/dictgen/src/rootcling_impl.cxx
Comment (hahnjo) [inline core/dictgen/src/rootcling_impl.cxx:62]: This looks unneeded

File: cmake/modules/SetUpFreeBSD.cmake
Comment (hahnjo) [inline cmake/modules/SetUpFreeBSD.cmake:12]: Is there really an Intel compiler for FreeBSD? I'd throw out anything that is not needed / tested on FreeBSD, probably only leave `freebsdadm64` as `ROOT_ARCHITECTURE`

File: cmake/modules/SetUpFreeBSD.cmake
Comment (hahnjo) [inline cmake/modules/SetUpFreeBSD.cmake:63]: `lld` is the default on FreeBSD anyway, no?

File: cmake/modules/SetUpFreeBSD.cmake
Comment (hahnjo) [inline cmake/modules/SetUpFreeBSD.cmake:114]: same here

File: core/unix/CMakeLists.txt
Comment (hahnjo) [reply core/unix/CMakeLists.txt]: same for all other CMake code as well :wink:

File: interpreter/cling/include/cling/libc.modulemap
Comment (hahnjo) [reply interpreter/cling/include/cling/libc.modulemap:73]: I think we should just remove this module: https://github.com/root-project/root/pull/12999 (assuming this doesn't cause more problems for FreeBSD...)

Comment (vgvassilev) [reply interpreter/cling/include/cling/libc.modulemap:73]: @mkrzewic, now I understand. Thanks @hahnjo for your PR! I've accepted it.

File: cmake/modules/SetUpFreeBSD.cmake
Comment (mkrzewic) [reply cmake/modules/SetUpFreeBSD.cmake:63]: indeed, cleaning up

File: cmake/modules/SetUpFreeBSD.cmake
Comment (mkrzewic) [reply cmake/modules/SetUpFreeBSD.cmake:12]: I got a request from the port maintainer (@eamjensen) to keep (and add/fix some of the FBSD specific naming) the cpu architectures in the SetupFreeBSD. Maybe refs to the old freebsd archs can be removed (freebsd4,freebsd5,freebsd7 in config/root-config.in and a bunch of options in etc/Makefile.arch) - instead we could glob/wildcard as they all seem to to the same thing for now.
I'll remove the intel thing, it makes no sense indeed.

Comment (mkrzewic) [reply cmake/modules/SetUpFreeBSD.cmake:12]: I cleaned up a bit, all in new commits.

File: core/dictgen/src/rootcling_impl.cxx
Comment (jalopezg-git) [inline core/dictgen/src/rootcling_impl.cxx:218]: I think we should be consistent on the use of `defined(R__FBSD)` and `defined(__FreeBSD__)`.  IMHO, we should stick to `R__FBSD` except perhaps in cling, which can be build standalone, where we should use `defined(__FreeBSD__)` instead.
```suggestion
#if defined(R__FBSD)
```

File: core/foundation/src/FoundationUtils.cxx
Comment (jalopezg-git) [inline core/foundation/src/FoundationUtils.cxx:138]: I see that this part is common with the `WIN32` case.  Perhaps we can restructure this a bit to avoid code duplication.

File: core/dictgen/src/rootcling_impl.cxx
Comment (jalopezg-git) [inline core/dictgen/src/rootcling_impl.cxx:220]: I saw some other usages of the procfs filesystem typically mounted on `/proc` on Linux, namely in:
- The `proof/`, `rootx/` and `bindings/pyroot/cppyy/` directories
- The `config/thisroot.sh` file

Maybe we should take a look into those too.

File: core/metacling/src/rootclingTCling.cxx
Comment (jalopezg-git) [inline core/metacling/src/rootclingTCling.cxx:25]: These are `extern`, no?

File: core/unix/src/TUnixSystem.cxx
Comment (jalopezg-git) [inline core/unix/src/TUnixSystem.cxx:423]:
```suggestion
      procstat* ps = procstat_open_sysctl();
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (jalopezg-git) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:47]: Interesting :monocle_face:...

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (jalopezg-git) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:1387]:
```suggestion
      procstat* ps = procstat_open_sysctl();
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (jalopezg-git) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:59]:
```suggestion
#endif
```

File: io/io/src/TFile.cxx
Comment (jalopezg-git) [inline io/io/src/TFile.cxx:173]:
```suggestion
#define getxattr(path, name, value, size) extattr_get_file(path, EXTATTR_NAMESPACE_USER, name, value, size)
```

File: cmake/modules/SetUpFreeBSD.cmake
Comment (jalopezg-git) [inline cmake/modules/SetUpFreeBSD.cmake:1]:
```suggestion
# Copyright (C) 1995-2023, Rene Brun and Fons Rademakers.
```

File: core/foundation/src/FoundationUtils.cxx
Comment (mkrzewic) [reply core/foundation/src/FoundationUtils.cxx:138]: possibly, I even tried, but for some reason this was better if I remember correctly... maybe not a good reson though:)

File: core/metacling/src/rootclingTCling.cxx
Comment (mkrzewic) [reply core/metacling/src/rootclingTCling.cxx:25]: possibly, this is a "legacy" fix from the port maintainer, it is needed and sufficient (it seems). I think maybe for some reason the symbols needed to be visible for lld, possibly some subtle difference with ld.

File: cmake/modules/SetUpFreeBSD.cmake
Comment (mkrzewic) [reply cmake/modules/SetUpFreeBSD.cmake:1]: well, this file is not quite copyright Rene and Fons if we are being strict:) But let's go with it.

Comment (Axel-Naumann) [reply cmake/modules/SetUpFreeBSD.cmake]: Appreciated; this is something we're working on addressing.

**Review by Axel-Naumann — APPROVED** [review]

File: core/dictgen/src/rootcling_impl.cxx
Comment (mkrzewic) [reply core/dictgen/src/rootcling_impl.cxx:220]: config/thisroot.sh should work as it is
rootx I'll fix, but is there no common library/place to define GetExePath (and the like) once for all components?
proof and python I'll leave for later as I have no use for those. Someone will contribute a patch if there is need:)

**Comment by mkrzewic** [comment]
I have one more question, this gets it building and kinda working, but there are still runtime issues: bunch of tests fail, most notably some io (basic works, higher compression levels fail) and threading (some TThread tests fail, for various reasons, from trivial _REENTRANT macro not defined, to proper crashes). Question is: keep adding to this PR or open a new issue some time in the future?

**Comment by Axel-Naumann** [comment]
New issues, new PRs please - we prefer to work incrementally, grandiose things have the habit to fail :-) I very much appreciate this first step to make it build!

**Comment by hahnjo** [comment]
Let's not downplay the achievement here - my understanding is that this PR not only allows to compile ROOT on FreeBSD, but gets us a decently functional executable for first tests and further improvements :rocket: I agree with Axel that other problems and tests should be addressed in future issues / PRs.

I think the final point to consider is how to merge. With the current state of commits addressing review comments, I would propose to squash on merge; unless @mkrzewic wants to rework the commits and fold the followups into the respective commits :wink:

**Comment by Axel-Naumann** [comment]
Thanks @hahnjo - @mkrzewic unless you have any objection I will merge this tomorrow in a squashed version.

**Comment by mkrzewic** [comment]
@Axel-Naumann sure, squash away, then I'll rebase my next branch and open a new PR.

**Comment by Axel-Naumann** [comment]
Thanks a LOT for your work, @mkrzewic !

**Comment by hahnjo** [comment]
We forgot to run the old Jenkins CI for the latest push (that is on us), https://github.com/root-project/root/pull/13073 takes care of an `unused-but-set-variable` warning...

**Pull request merged**
